src/symbol/collision_feature.js

```diff
@@ -72,42 +72,60 @@ class CollisionFeature {
     _addLineCollisionBoxes(collisionBoxArray, line, anchor, segment, labelLength, boxSize, featureIndex, sourceLayerIndex, bucketIndex) {
         const step = boxSize / 2;
         const nBoxes = Math.floor(labelLength / step);
+        // We calculate line collision boxes out to 150% of what would normally be our
+        // max size, to allow collision detection to work on labels that expand as
+        // they move into the distance
+        const nPitchPaddingBoxes = Math.floor(nBoxes / 4);
```

Comment:
I pushed a commit that skips adding these boxes to the CollisionTile on unpitched maps, so at least in that case there shouldn't be much of a performance impact. There is probably a more optimal version in which we calculate the number of padding boxes based on the tile distance, but I'm not worrying about that for now.